### PR TITLE
Barebones Scouting

### PR DESCRIPTION
**Pull request by imann24**
***Disclaimer**: This PR is still very much in progress, but I've changed so much in it that it seems prudent to submit it for an initial review and feedback before I go any farther*

**Scripts**
- Parsing Dog Sprites from Sprites Sheets
- Added Scouting Mechanic (mostly)
- Added Concept of Real Time and Last Time Saved into Serialization
- Added Tons of Delegates (too many to list) across UI and data scripts
- Added GiftController for handling reward logic from dogs scouting
- And much, much more

**Scenes**
- Added test scene for scouting report display
- Made modifications to Home to support scouting

**Sprites**
- Added three new spritesheets of dogs

**JSON**
- Added new JSON for spritesheets
- Edited Tuning JSON to include more vars to relate to gifts from scouting
- Updated the GameData JSON w/ New Breed / Dog / Sprite info

### CODE REVIEW

File: Assets/Scripts/UI/Pickup Pup/Currency/CurrencyDisplay.cs [hunk 1]
```diff
@@ -5,19 +5,68 @@
...
-
-    // Updates text to show the new amount of Currency
-    public void OnUpdate()
-    {
-        text.text = currency.Amount.ToString();
-    }    
+	CurrencyType type;
+	PPDataController dataController;
+
+	public void Init(PPDataController dataController, CurrencyType type)
```
Comment (gracebarsny) [inline Assets/Scripts/UI/Pickup Pup/Currency/CurrencyDisplay.cs]: Why does this take in a PPDataController?

File: Assets/Scripts/UI/Pickup Pup/DogSlot/DogOutsideSlot.cs [hunk 2]
```diff
@@ -31,22 +75,62 @@ public override void Init(Dog dog)
...
+		this.dog = dog;
+		this.dogInfo = dog.Info;
+		subscribeTimerEvents(dog);
+		nameText.text = dog.Name;
+		dogImage.sprite = dog.Portrait;
+		dog.SetTimer(dogInfo.TimeRemainingScouting);
+		dog.ResumeTimer();
+	}
+		
+	public override void ClearSlot ()
```
Comment (gracebarsny) [inline Assets/Scripts/UI/Pickup Pup/DogSlot/DogOutsideSlot.cs]: Remove space between ClearSlot and (

File: Assets/Scripts/UI/Pickup Pup/DogSlot/DogOutsideSlot.cs [hunk 3]
```diff
@@ -31,22 +75,62 @@ public override void Init(Dog dog)
...
+	{
+		// Call Dog functionality first because base method sets dog ref to null:
+		dog.StopTimer();
+		dogImage.sprite = collarSprite;
+		nameText.text = string.Empty;
+		timerText.text = string.Empty;
+		base.ClearSlot ();
```
Comment (gracebarsny) [inline Assets/Scripts/UI/Pickup Pup/DogSlot/DogOutsideSlot.cs]: Remove space between ClearSlot and (

File: Assets/Scripts/UI/Pickup Pup/ScoutingDisplay/ScoutingDisplay.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+			slot.ResumeScouting(dog);
+		}
+	}
+
+	public void SubscribeToTimerEnd(Dog dog)
+	{
+		dog.SubscribeToScoutingTimerEnd(handleScoutingTimerEnd);
+	}
+
+	#region Override MonoBehaviourExtended
```
Comment (gracebarsny) [inline Assets/Scripts/UI/Pickup Pup/ScoutingDisplay/ScoutingDisplay.cs]: I believe we decided on "MonoBehaviourExtended Overrides," no?

File: Assets/Scripts/UI/Pickup Pup/ScoutingDisplay/ScoutingDisplay.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+		DogOutsideSlot slot;
+		if(slotsByIndex.TryGetValue(slotIndex, out slot))
+		{
+			slot.ResumeScouting(dog);
+		}
+	}
+
+	public void SubscribeToTimerEnd(Dog dog)
+	{
+		dog.SubscribeToScoutingTimerEnd(handleScoutingTimerEnd);
```
Comment (gracebarsny) [inline Assets/Scripts/UI/Pickup Pup/ScoutingDisplay/ScoutingDisplay.cs]: Would be nice if we could use MonoBehaviourExtended's subscribeEvents method, but I suppose that might not be possible.

File: Assets/Scripts/UI/Pickup Pup/ShopScreen/PPShopUIController.cs [hunk 6]
```diff
@@ -36,8 +36,9 @@ protected override void fetchReferences()
         EventController.Event(PPEvent.LoadShop);
 
         // Set Currency Displays
-        dogFoodDisplay.SetCurrency(gameController.DogFood);
-        coinDisplay.SetCurrency(gameController.Coins);
+		// Set Currency Displays
```
Comment (gracebarsny) [inline Assets/Scripts/UI/Pickup Pup/ShopScreen/PPShopUIController.cs]: I think you copied this comment twice.

File: Assets/Scripts/Util/FileUtil.cs [hunk 7]
```diff
@@ -9,6 +9,20 @@
 
 public static class FileUtil 
 {
+	public static string CreatePath (params string[] directories) {
```
Comment (gracebarsny) [inline Assets/Scripts/Util/FileUtil.cs]: Remove space between CreatePath and (
Move bracket onto its own line

File: Assets/Scripts/Util/FileUtil.cs [hunk 8]
```diff
@@ -9,6 +9,20 @@
...
+			string path = Path.Combine(directories[0], directories[1]);
+			for (int i = 2; i < directories.Length; i++) {
+				path = Path.Combine(path, directories[i]);
+			}
+			return path;
+		} else if (directories.Length == 1) {
+			return Path.Combine(directories[0], string.Empty);
+		} else {
+			return Path.Combine(string.Empty, string.Empty);
+		}
```
Comment (gracebarsny) [inline Assets/Scripts/Util/FileUtil.cs]: BRACKETS

File: Assets/Scripts/UI/Pickup Pup/Currency/CurrencyDisplay.cs [hunk 9]
```diff
@@ -5,19 +5,68 @@
...
+			case CurrencyType.DogFood:
+				updateAmount(dataController.DogFood.Amount);
+				break;
+		}
+	}
+
+
+	#region MonoBehaviourExtended Overrides
+
+	protected override void subscribeEvents ()
```
Comment (gracebarsny) [inline Assets/Scripts/UI/Pickup Pup/Currency/CurrencyDisplay.cs]: Remove space between subscribeEvents and (

File: Assets/Scripts/UI/Pickup Pup/Currency/CurrencyDisplay.cs [hunk 10]
```diff
@@ -5,19 +5,68 @@
...
+	{
+		base.subscribeEvents ();
```
Comment (gracebarsny) [inline Assets/Scripts/UI/Pickup Pup/Currency/CurrencyDisplay.cs]: Remove space before (

File: Assets/Scenes/Shop.unity [hunk 11]
```diff
@@ -1662,6 +1662,11 @@ Prefab:
       propertyPath: m_RootOrder
       value: 3
       objectReference: {fileID: 0}
+    - target: {fileID: 114426272054950630, guid: 6c18c567384a5444e8446d670695ec79,
+        type: 2}
+      propertyPath: tryAlsoCallStringsAsPPEvents
+      value: 0
+      objectReference: {fileID: 0}
```
Comment (gracebarsny) [inline Assets/Scenes/Shop.unity]: Did you mean to modify this scene?

File: Assets/Scripts/Data/Pickup Pup/PPGameSave.cs [hunk 12]
```diff
@@ -24,14 +32,53 @@ public CurrencySystem Currencies
...
 	#endregion
 
-    CurrencySystem currencies;
-
-	public PPGameSave(DogDescriptor[] dogs, CurrencySystem currencies) 
+	public PPGameSave(DogDescriptor[] adoptedDogs, DogDescriptor[] scoutingDogs, CurrencySystem currencies) 
 	{
-		this.AdoptedDogs = new List<DogDescriptor>(dogs);
+		this.AdoptedDogs = new List<DogDescriptor>(adoptedDogs);
+		this.ScoutingDogs = new List<DogDescriptor>(scoutingDogs);
```
Comment (imann24) [inline Assets/Scripts/Data/Pickup Pup/PPGameSave.cs]: Perhaps move in to `DogDatabse`? Or make a new class: `DogStorageSystem`? Either way, will hold off un such refactors until after this is merged and we're past first playable

File: Assets/Scripts/UI/Pickup Pup/ScoutingDisplay/ScoutingDisplay.cs (hunk 5 above)
Comment (imann24) [inline Assets/Scripts/UI/Pickup Pup/ScoutingDisplay/ScoutingDisplay.cs]: Yeah, the parent method is more for subscriptions on init. This one is meant to be used more dynamically

File: Assets/Scenes/Shop.unity (hunk 11 above)
Comment (imann24) [inline Assets/Scenes/Shop.unity]: I think was a modification from a fix to EventController

**Comment by imann24** [comment]
@gracebarsny Do you want me to make the change for saving more frequently in this PR or in a future one?

File: Assets/Scripts/Data/Pickup Pup/PPGameSave.cs (hunk 12 above)
Comment (gracebarsny) [inline Assets/Scripts/Data/Pickup Pup/PPGameSave.cs]: Of course. Maybe even `DogManagement`.

File: Assets/Scripts/UI/Pickup Pup/Currency/CurrencyDisplay.cs (hunk 1 above)
Comment (gracebarsny) [inline Assets/Scripts/UI/Pickup Pup/Currency/CurrencyDisplay.cs]: So you can't use `PPDataController.GetInstance`?

**Comment by gracebarsny** [comment]
@imann24 Additional saves can go in a future PR.

File: Assets/Scripts/UI/Pickup Pup/Currency/CurrencyDisplay.cs (hunk 1 above)
Comment (imann24) [inline Assets/Scripts/UI/Pickup Pup/Currency/CurrencyDisplay.cs]: Yes, but I like use Dependency Injection when possible (https://en.wikipedia.org/wiki/Dependency_injection). I think it makes the code cleaner

File: Assets/Scripts/Data/Pickup Pup/PPGameSave.cs (hunk 12 above)
Comment (imann24) [inline Assets/Scripts/Data/Pickup Pup/PPGameSave.cs]: Yours is my favorite so far!

**Pull request merged**